### PR TITLE
Interoperable DNS Server Cookies support

### PR DESCRIPTION
**Pull request by wtoorop**

### CODE REVIEW

**Comment by wtoorop** [comment]
Copy TODO from @magenbluten's branch:

- [x] unit tests

File: server.c [hunk 1]
```diff
@@ -3580,6 +3583,7 @@ handle_tcp_reading(int fd, short event, void* arg)
 	ssize_t received;
 	struct event_base* ev_base;
 	struct timeval timeout;
+	uint32_t now;
```
Comment (wcawijngaards) [inline server.c]:
```suggestion
	uint32_t now = 0;
```
Perhaps this is useful

File: util.c [hunk 2]
```diff
@@ -1234,3 +1235,58 @@ int set_cpu_affinity(cpuset_t *set)
...
+{
+	/* New cookie secret becomes the staging secret (position 1)
+	 * unless there is no active cookie yet, then it becomes the active
+	 * secret.  If the NSD_COOKIE_HISTORY_SIZE > 2 then all staging cookies
+	 * are moved one position down.
+	 */
+	if(nsd->cookie_count == 0) {
+		memcpy( nsd->cookie_secrets->cookie_secret
+		       , secret, NSD_COOKIE_SECRET_SIZE);
+		nsd->cookie_count = 1;
```
Comment (wcawijngaards) [inline util.c:1249]:
```suggestion
		nsd->cookie_count = 1;
		memset(secret, 0, NSD_COOKIE_SECRET_SIZE);
```
The erase also happens later, so here too.

File: util.c [hunk 3]
```diff
@@ -1234,3 +1235,58 @@ int set_cpu_affinity(cpuset_t *set)
...
+	 */
+	if(nsd->cookie_count < 2)
+		return;
+	memcpy( active_secret, nsd->cookie_secrets[0].cookie_secret
+	      , NSD_COOKIE_SECRET_SIZE);
+	memmove( &nsd->cookie_secrets[0], &nsd->cookie_secrets[1]
+	       , sizeof(struct cookie_secret) * (NSD_COOKIE_HISTORY_SIZE - 1));
+	memcpy( nsd->cookie_secrets[nsd->cookie_count - 1].cookie_secret
+	      , active_secret, NSD_COOKIE_SECRET_SIZE);
+	memset(active_secret, 0, NSD_COOKIE_SECRET_SIZE);
```
Comment (wcawijngaards) [inline util.c]: These kinds of zero memsets get removed by the compiler. Use `explicit_bzero` to make sure it does not get removed. Also for the other memset operations on the cookies.

File: remote.c [hunk 4]
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
...
+	for(size_t i = 0; i < nsd->cookie_count; i++) {
+		struct cookie_secret const* cs = &nsd->cookie_secrets[i];
+		ssize_t const len = hex_ntop(cs->cookie_secret, NSD_COOKIE_SECRET_SIZE,
+																 secret_hex, sizeof(secret_hex));
+		(void)len; /* silence unused variable warning with -DNDEBUG */
+		assert( len == NSD_COOKIE_SECRET_SIZE * 2 );
+		secret_hex[NSD_COOKIE_SECRET_SIZE * 2] = '\0';
+		fprintf(f, "%s\n", secret_hex);
+	}
+	fclose(f);
```
Comment (wcawijngaards) [inline remote.c:2190]: If you want to memset the array, the 'secret_hex' array needs to be cleared here.

File: remote.c [hunk 5]
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
...
+	if(strlen(arg) != 32) {
+		(void)ssl_printf(ssl, "invalid cookie secret: invalid argument length\n");
+		(void)ssl_printf(ssl, "please provide a 128bit hex encoded secret\n");
+		memset(arg, 0xdd, strlen(arg));
+		return;
+	}
+	if(hex_pton(arg, secret, NSD_COOKIE_SECRET_SIZE) != NSD_COOKIE_SECRET_SIZE ) {
+		(void)ssl_printf(ssl, "invalid cookie secret: parse error\n");
+		(void)ssl_printf(ssl, "please provide a 128bit hex encoded secret\n");
+		memset(arg, 0xdd, strlen(arg));
```
Comment (wcawijngaards) [inline remote.c]: If you memset 'arg' you should also clear the 'secret' array, which contains partial results. Also for memset calls for arg further in this function, there should be a clear of the 'secret' array, as well, if you want to erase parts of that data.

File: remote.c [hunk 6]
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
...
+		(void)len; /* silence unused variable warning with -DNDEBUG */
+		assert( len == NSD_COOKIE_SECRET_SIZE * 2 );
+		secret_hex[NSD_COOKIE_SECRET_SIZE * 2] = '\0';
+		if (i == 0)
+			(void)ssl_printf(ssl, "active : %s\n",  secret_hex);
+		else if (nsd->cookie_count == 2)
+			(void)ssl_printf(ssl, "staging: %s\n",  secret_hex);
+		else
+			(void)ssl_printf(ssl, "staging[%d]: %s\n", i, secret_hex);
+	}
```
Comment (wcawijngaards) [inline remote.c:2312]: If you want to clear the array, here `secret_hex` is the array to clear.

File: remote.c [hunk 7]
```diff
@@ -2226,6 +2370,14 @@ execute_cmd(struct daemon_remote* rc, RES* ssl, char* cmd, struct rc_state* rs)
 		do_assoc_tsig(ssl, rc->xfrd, skipwhite(p+10));
 	} else if(cmdcmp(p, "del_tsig", 8)) {
 		do_del_tsig(ssl, rc->xfrd, skipwhite(p+8));
+	} else if(cmdcmp(p, "add_cookie_secret", 17)) {
+		do_add_cookie_secret(ssl, rc->xfrd, skipwhite(p+18));
```
Comment (wcawijngaards) [inline remote.c]:
```suggestion
		do_add_cookie_secret(ssl, rc->xfrd, skipwhite(p+17));
```
The string length is 17.

File: remote.c [hunk 8]
```diff
@@ -2226,6 +2370,14 @@ execute_cmd(struct daemon_remote* rc, RES* ssl, char* cmd, struct rc_state* rs)
...
+	} else if(cmdcmp(p, "drop_cookie_secret", 18)) {
+		do_drop_cookie_secret(ssl, rc->xfrd, skipwhite(p+18));
+	} else if(cmdcmp(p, "print_cookie_secrets", 20)) {
+		do_print_cookie_secrets(ssl, rc->xfrd, skipwhite(p+20));
+	} else if(cmdcmp(p, "activate_cookie_secret", 22)) {
+		do_activate_cookie_secret(ssl, rc->xfrd, skipwhite(p+18));
```
Comment (wcawijngaards) [inline remote.c]:
```suggestion
		do_activate_cookie_secret(ssl, rc->xfrd, skipwhite(p+22));
```
The string length is 22.

File: remote.c [hunk 9]
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
...
+	/* open write only and truncate */
+	if((f = fopen(secret_file, "w")) == NULL ) {
+		(void)ssl_printf(ssl, "unable to open cookie secret file %s: %s",
+		                 secret_file, strerror(errno));
+		return 0;
+	}
+	for(size_t i = 0; i < nsd->cookie_count; i++) {
+		struct cookie_secret const* cs = &nsd->cookie_secrets[i];
+		ssize_t const len = hex_ntop(cs->cookie_secret, NSD_COOKIE_SECRET_SIZE,
+																 secret_hex, sizeof(secret_hex));
```
Comment (wcawijngaards) [inline remote.c]:
```suggestion
			secret_hex, sizeof(secret_hex));
```
Indent, nothing important.

**Review by wcawijngaards — APPROVED** [review]
Code looks good, thanks for already incorporating a bunch of changes!

File: remote.c (hunk 4 above)
Comment (wtoorop) [reply remote.c:2190]: resolved in aa3bca3

File: remote.c (hunk 5 above)
Comment (wtoorop) [reply remote.c]: resolved in aa3bca3

File: remote.c (hunk 6 above)
Comment (wtoorop) [reply remote.c:2312]: resolved in aa3bca3

**Comment by wtoorop** [comment]
Thanks @wcawijngaards ! Maybe you'd like to rereview the last commit?

**Pull request merged**